### ISSUE TITLE
Creating a cluster with kubeadm

### ISSUE THREAD

**Issue by Alestrix** [kubernetes/website#37285]
The page [Creating a cluster with kubeadm](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/) references the documentation on container runtime installation (https://kubernetes.io/docs/setup/production-environment/container-runtimes/). There the information is given that for systemd Linuxes, the `cgroupDriver` must be set in a **configuration file** that is passed to `kubeadm` via `--config`.

The documentation in [Creating a cluster with kubeadm](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/) explains that it is or might be necessary to pass various parameters to `kubeadm init` like `--pod-network-cidr` or `--control-plane-endpoint`. However, this does not work together with the aforementioned `--config` parameter.

Also, regarding the recommendation to use `--control-plane-endpoint` for possible future HA setups, the documentation is missing the info how to add the HA DNS name of the control node to the x509's SAN fields. If that's missing, no communication to the control node will be possible. (in my case, the control nodes main hostname was k8s-control01, but for future-proving, I aliased it to k8s-control).

Then, there's more: After following the directions to install docker engine and cri-dockerd, I end up with two container runtime control sockets, so I had to add `--cri-socket` parameter's equivalent to the config file to explicitly point to one of the sockets.

After lots of googling I finally had all the little pieces together:

```
kind: ClusterConfiguration
apiVersion: kubeadm.k8s.io/v1beta3
kubernetesVersion: v1.25.2
controlPointEndpoint: "k8s-control:6443"
networking:
  podSubnet: 10.69.0.0/16
apiServer:
  certSANs:
  - k8s-control
---
kind: KubeletConfiguration
apiVersion: kubelet.config.k8s.io/v1beta1
cgroupDriver: systemd
---
apiVersion: kubeadm.k8s.io/v1beta3
kind: InitConfiguration
nodeRegistration:
  criSocket: "unix:///run/cri-dockerd.sock"
---
apiVersion: kubeadm.k8s.io/v1beta3
kind: JoinConfiguration
nodeRegistration:
  name: k8s-control
```

TL;DR: Please add the config.yaml equivalents of the various `--some-setting` kubeadm parameters.

**Comment by neolit123** [comment]
/sig cluster-lifecycle

thank you for the feedback.
from time to time we do get complains about how the documentation is organized, yet we need clarity of what has to be changed exactly. if the documentation does not work for someone we usually ask them to draft a pull request and see if we can implement the changes they have in mind.

> The page [Creating a cluster with kubeadm](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/) references the documentation on container runtime installation (https://kubernetes.io/docs/setup/production-environment/container-runtimes/). There the information is given that for systemd Linuxes, the cgroupDriver must be set in a configuration file that is passed to kubeadm via --config.

as long as you follow the CR page guide and the "kubeadm create a cluster" guide, the driver will be "systemd" everywhere as required by kubeadm.

> The documentation in [Creating a cluster with kubeadm](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/) explains that it is or might be necessary to pass various parameters to kubeadm init like --pod-network-cidr or --control-plane-endpoint. However, this does not work together with the aforementioned --config parameter.

that is true, flags are mutually exclusive with --config options. but all options that are flags are present as config options.
they can be found in the API page that is linked.

> Also, regarding the recommendation to use --control-plane-endpoint for possible future HA setups, the documentation is missing the info how to add the HA DNS name of the control node to the x509's SAN fields. If that's missing, no communication to the control node will be possible. (in my case, the control nodes main hostname was k8s-control01, but for future-proving, I aliased it to k8s-control).

the CPE DNS name is automatically added in the cert SANs.

> Then, there's more: After following the directions to install docker engine and cri-dockerd, I end up with two container runtime control sockets, so I had to add --cri-socket parameter's equivalent to the config file to explicitly point to one of the sockets.

containerd / crio are more recommend than cri-dockerd + docker engine, since the former projects are more actively maintained.
yes, you do need to be explicit about the socket if multiple CRs are present on the host.

> kind: KubeletConfiguration
apiVersion: kubelet.config.k8s.io/v1beta1
cgroupDriver: systemd

this is not required, because the latest versions of kubeadm default the to kubeletconfiguration to "systemd". that can be found in link in a (not so well titled) section on CR page:
https://kubernetes.io/docs/setup/production-environment/container-runtimes/#migrating-to-the-systemd-driver-in-kubeadm-managed-clusters

**Comment by sftim** [comment]
> CPE

Customer premises equipment?
